### PR TITLE
Don't call RuntimeHelper.Equals in Object.Equals

### PR DESCRIPTION
**Pull request by huoyaoyuan**
`RuntimeHelper.Equals` only deals value types with additional comparison, which would be handled by the override `ValueType.Equals`. The behavior of reference types is just `ReferenceEquals`.

Question: `RuntimeHelper.Equals` compares value types for value and padding, while `ValueType.Equals` (fast path) compares only fields. I looked for the purpose of `RuntimeHelper.Equals`, and found it was introduced in .NET Framework 1.1. Shouldn't padding just be an implementation detail, and the purpose of the API is comparing value bits?

Should we unify the implementation with `ValueType.Equals` then? The Mono behavior is actually wrong now (uses the full path of `ValueType.Equals`, instead of fast compare bits path).

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
`RuntimeHelper.Equals` just forbids GC and compares raw references, oh...

File: src/libraries/System.Private.CoreLib/src/System/Object.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public Object()
         /// <returns><c>true</c> if the specified object is equal to the current object; otherwise, <c>false</c>.</returns>
         public virtual bool Equals(object? obj)
         {
-            return RuntimeHelpers.Equals(this, obj);
+            return ReferenceEquals(this, obj);
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Object.cs:50]: can be just `return this == obj;` I guess (shorter and saves time for inliner)

Also I assume you can delete the `RuntimeHelpers.Eqauls` FCall now

**Comment by EgorBo** [comment]
> Question: RuntimeHelper.Equals compares value types for value and padding

the problem is not only in paddings (which might contain garbage values) but also with e.g. floats - `RuntimeHelpers.Equals(float.NaN, float.NaN)` currently returns `true` which is not correct. It also, I guess, compares gc refs by pointers. Not sure anyone actually uses this method and it probably can be just deprecated (the docs don't mention the pit falls)

File: src/libraries/System.Private.CoreLib/src/System/Object.cs (hunk 1 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Object.cs:50]: Actually you can rewrite it to C# and delete the FCall, thus, ILLink will be able to trim it out if it's unused.

**Comment by huoyaoyuan** [comment]
> `RuntimeHelpers.Equals(float.NaN, float.NaN)` currently returns `true` which is not correct.

`double.NaN.Equals(double.NaN)` returns true, different from `==`. The difference of `Equals` and `==` can be misleading here.

I guess the purpose for it should be comparing fields by bitwise equality, which is reasonable for value type.

The FCall implementation also handles string equality, but never invoked because guarded by `IsValueType`.

Maybe some elders like Jan can tell its history.

**Comment by jkotas** [comment]
`RuntimeHelpers.Equals` is a broken API. https://github.com/dotnet/runtime/issues/19379#issuecomment-263101741 . Nobody should be using. The only change we may want to make is to mark it as obsolete.

**Comment by huoyaoyuan** [comment]
So for simplification, just delegate the call to `ValueType.Equals` is sufficient?

File: src/libraries/System.Private.CoreLib/src/System/Object.cs (hunk 1 above)
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Object.cs:50]:
```suggestion
            return this == obj;
```

**Comment by jkotas** [comment]
> So for simplification, just delegate the call to ValueType.Equals is sufficient?

It would change the behavior. There is no good reason for trying to make breaking changes in `RuntimeHelpers.Equals` behavior.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**